### PR TITLE
Fix progress bar interactions in `caliban run` subprocesses

### PR DESCRIPTION
**Pull request by sritchie**
This PR makes a few changes that allow us to use `caliban run` to execute scripts with (single) progress bars or other updating progress meters that use carriage returns to modify lines.

It's not PERFECT — if you have a progress bar and use `tqdm.write("HI!")` inside, for example, you'll trigger a newline, instead of the nice, flowing-upward output above the inner progress bar that you get when you run `tqdm` locally. But that is a challenge for another day.

I was able to package this pretty cleanly, I think, into `TqdmFile`, but it does require some fiddling in `capture_output` as well.

### CODE REVIEW

File: caliban/docker.py [hunk 1]
```diff
@@ -32,15 +32,15 @@
 import tqdm
 from absl import logging
 from blessings import Terminal
+from tqdm.utils import _screen_shape_wrapper
 
 import caliban.config as c
 import caliban.util as u
-from caliban.history.utils import (get_sql_engine, get_mem_engine,
```
Comment (sritchie) [inline caliban/docker.py:38]: import sorting, no changes.

File: caliban/util.py [hunk 2]
```diff
@@ -402,19 +403,23 @@ def capture_stdout(cmd: List[str],
...
   with subprocess.Popen(cmd,
                         stdin=subprocess.PIPE,
                         stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT,
-                        bufsize=1,
-                        encoding="utf-8") as p:
+                        universal_newlines=False,
+                        bufsize=1) as p:
     if input_str:
-      p.stdin.write(input_str)
```
Comment (sritchie) [inline caliban/util.py:412]: TextIOWrapper below doesn't like the encoding, so I removed it from the constructor above and manually apply it here on stdin.

File: caliban/util.py [hunk 3]
```diff
@@ -402,19 +403,23 @@ def capture_stdout(cmd: List[str],
...
+      p.stdin.write(input_str.encode('utf-8'))
     p.stdin.close()
 
-    for line in p.stdout:
-      print(line, end='', file=file)
+    out = io.TextIOWrapper(p.stdout, newline='')
```
Comment (sritchie) [inline caliban/util.py:417]: this method preserves carriage returns!

File: caliban/util.py [hunk 4]
```diff
@@ -671,13 +676,22 @@ def validated_file(path: str) -> str:
...
   """Dummy file-like that will write to tqdm"""
   file = None
+  prefix = _term_move_up() + '\r'
 
   def __init__(self, file):
     self.file = file
+    self._carriage_pending = False
+
+  def write(self, line):
+    if self._carriage_pending:
```
Comment (sritchie) [inline caliban/util.py:686]: tqdm is a little busted; you can't directly pass carriage returns in, so I put some logic here that appends the required prefix to the NEXT line, and swaps the carriage return with a newline.

File: setup.py [hunk 5]
```diff
@@ -43,7 +43,7 @@ def readme():
     'commentjson',
     'google-api-python-client',
     'pyyaml',
-    'tqdm',
+    'tqdm>=4.45.0',
```
Comment (sritchie) [inline setup.py:46]: the utility function requires a pinned version.

File: tutorials/basic/mnist.py [hunk 6]
```diff
@@ -58,7 +60,7 @@ def main(_):
   print(
       f'Training model with learning rate={FLAGS.learning_rate} for {FLAGS.epochs} epochs.'
   )
-  model.fit(x_train, y_train, epochs=FLAGS.epochs, verbose=2)
+  model.fit(x_train, y_train, epochs=FLAGS.epochs)
```
Comment (sritchie) [inline tutorials/basic/mnist.py:63]: now we can allow the progress bars to shine through!

File: caliban/docker.py (hunk 1 above)
Comment (ajslone) [reply caliban/docker.py:38]: ack

File: caliban/util.py (hunk 4 above)
Comment (ajslone) [reply caliban/util.py:686]: nice, that's a tricky fix there

File: caliban/util.py (hunk 3 above)
Comment (ajslone) [reply caliban/util.py:417]: another nice solution

File: tutorials/basic/mnist.py (hunk 6 above)
Comment (ajslone) [reply tutorials/basic/mnist.py:63]: vary nice

**Pull request merged**